homepage/templates/homepage/app_layout.html

```diff
@@ -48,6 +54,24 @@
   </div>  
 </nav>
 
+
+<style>
+  .messages {
+    background-color:dodgerblue;
+    color: white;
+    padding: 10px;
+  }
+  </style>
+  
+  {% if messages %}
+  <ul class="messages">
+      {% for message in messages %}
+      <li{% if message.tags %} class="{{ message.tags }}"{% endif %}>{{ message }}</li>
+      {% endfor %}
+  </ul>
+  {% endif %}

```

Comment:
66 - 72:
when I talked about displaying error messages that are raised from the form (sing up, log in) I meant that they appear right above the form itself, as marked in the related .html file below.

Reply:
Also, I think it might be easier to use the bootstrap built in component Alert for this:
```
<div class="alert alert-warning alert-dismissible fade show" role="alert">
    ...
</div>
```

Reply:
@VeraTs 
Thanks for your detailed feedback.

There are a coupe of reasons i implemented the messages display in **app_layout.html**:
(which makes **the messages appear in both the sign-in and the signup form**)

1.  **Duplicate code:** since both the sign-in and sign-up pages need to use the exact same code to display the messages, it only makes sense to implement the code once in app_layout which they both extend, instead of writing the same code twice in 2 different places.

2. **The messages need to appear in other pages as well** and not only in sign_in.html and sign_up.html because after you complete the sign-in \ sign-up process - you are redirected to a different page
Hence - implementing lines 66-72 where you suggested (sign_in.html) will not work for messages that need to be redirected to other pages.
If implemented once in app_layout.html -  the message is displayed in the correct page even if redirected.



- **Regarding bootstrap alerts:**
The only change this will make **is to add more colors to the messages** (each message type has its own color):
https://www.ordinarycoders.com/blog/article/django-messages-framework

Currently all the success / error / info messages are displayed in a blue color. (And IMO it looks good)
Since this is only a basic implementation, and i already implemented it this way - i prefer to keep it like this.
Ofc, in my feature PR for the sign-in and sign-up pages i might add the bootstrap alerts for more colors, **but IMO this is not a high priority and there are more important features we should focus on first** (we only have 10 more days for the demo..)


**Example of the current messages:**

![image](https://user-images.githubusercontent.com/79373827/118115105-2f7acf80-b3f1-11eb-9ee2-2ab4845fc321.png)


![image](https://user-images.githubusercontent.com/79373827/118115141-3f92af00-b3f1-11eb-9b58-e3b492b0c20c.png)

Reply:
@LeonRabinovich I agree with Vera. These messages shouldn't be in **app_layout.html**, as they have nothing in common with other features.
On all websites, when the user logs in, the user just gets redirected to the desired page or landing page.
When the user logs out, the user gets redirected to the login page.
For example, the error message ```invalid user name``` should be right below the sign-in button, and not placed somewhere on the screen.

Reply:
@LeonRabinovich If someone tries to sign up or log in and is unsuccessful, are they redirected to a page that isn't just the same?
Since in my experience, in most websites, they are redirected (or don't even leave) the page for the sign up/ log in.

Reply:
@oriAdler 
Please read my comment again.

The messages are in app_layout.html because many of our pages need to use the messages, and all pages extend app_layout.html. Also in the future when you implement messages in view.py for any page - they will be displayed automatically because of the code inside app_layout.html
it makes no sense to duplicate the code and place it in 3-4 diffrent pages.


`On all websites, when the user logs in, the user just gets redirected to the desired page or landing page.
When the user logs out, the user gets redirected to the login page.`
A message should be displayed after a user logs-in \ logs-out - and this is exactly what i implemented.
this is not possible if you put the messages code only in sign_in.html.

Reply:
@VeraTs 

To make it clear, i will list all of  the scenarios for sign-in:


1. **Sign-in is successful:** redirect to the homepage. successful sign-in message is displayed after re-direction in the homepage. (this is possible because the homage extends app_layout.html)
 
2. **Sign-in fails:** stay on the same page and clear the form (or just re-direct to the sign-in page) and display sign-in fails message. (Redirection will probably not work without implementing the messages in app_layout.html)

3. **Sign-out:** Redirect to the homepage and display a message ("<UserName> signed out successfully").
To display the message after the re-direction, the messages need to be implemeneted in app_layout.html because the homepage extends app_layout.html

Reply:
@VeraTs @oriAdler 

In addition to the comments above, i have 1 more thing to add.
The purpose of this PR is to implement the sign-in and sign-out functionality in the simplest way possible with simple clear messages.

There are other ways to implement the messages, however currently I'm using the simplest solution (which works and has all the functionality we need IMO).

If we had more time (again the final demo is in 10 days), i could make the messages look better and displayed better, but please note we are on a tight schedule.
Does it matter that much if the error message is displayed in the button of the screen or the top of the screen? (for the above purpose)

I think its better to have a simple solution that works so we can work on other features that rely on the sign-in functionality.(for example the profile page)
If we have time - we can make make the UI look better, but i think this should not be the main goal right now.

Reply:
@LeonRabinovich As all pages inherit from the app layout, this page is supposed to be minimal as possible.
All other pages don't use these messages.
So I believe it's best if the messages section will be on the sign-in page.
About the schedule, this feature was supposed to be completed a long time ago.

Reply:
> @LeonRabinovich As all pages inherit from the app layout, this page is supposed to be minimal as possible.
> All other pages don't use these messages.
> So I believe it's best if the messages section will be on the sign-in page.
> About the schedule, this feature was supposed to be completed a long time ago.

@oriAdler all pages extend app_layout which is exactly what makes this possible. (Implement messages once - use in all pages **if and when needed**)

Currently homepage, signin, signout, signup and User profile which will be added use those messages, without duplicating the code like you are suggesting.

This pull request has been open for 6 days, and I see no reason to delay it even more when we have 10 days left. As I said, other features rely on this feature.

Reply:
To be more specific, the PR for this feature was opened 8 days ago #96.
This conversation is only from yesterday, so this is not the reason for the delay.
The homepage doesn't display these messages, and we don't have a sign-out page on the application.
Anyway, as we are a team, we should make decisions together.
At the moment, you are the only one in favor of it.

Reply:
@oriAdler
Again, the homepage displays those messages after redirection..
for example, after clicking on the sign-out **button** from any page, you are redirected to the homepage, and a message is displayed for a sucsesfull sign out.(on the home page)

Anyways, even if I disagree about the implementation, I agree that this should be a team decision.

Hence - I ask @VeraTs  and @ellaml to read this thread discussion and let me know if you think I need to change the Implemntation.

I will do as the team decides.

Reply:
@LeonRabinovich I understand your desire for a simpler approach to the goal of this PR, given our timeline.
I even agree with it, somewhat, given that you want to display success messages as well.
However, please do note that it would not be code duplication to display it in several divs instead of in the framing, since it would serve a different purpose in each. 
For example - as to your cases, it would not be needed if in the future, sign out would lead to a "goodbye" type page instead of the front page. In case of the successful sign up - to redirect to a future "Profile" page. In both cases there would be no need to display success messages, since the pages themselves would be sufficient, and then implementation of the message display in app_layout would be redundant.
Therefore, for now, I am willing to be satisfied with the way it is now, in the app_layout even if it isn't the best approach overall.

That said, in regards to the html tags (style, head) please see what I said above, or consider deleting them and using the bootstrap for styling, since using the bootstrap classes does not require further styling and would ignore the issue altogether.

Reply:
@VeraTs 
I see your point, thank you for being satisfied with this approach for now.
Regarding the HTML tags - i will fix it and comment on the thread above when its commited.


@ellaml - i would appreciate your review on this PR as well, would help me to get to the profile page PR sooner.

--
I've also added a video in the original PR comment that showcases the entire sign-in, sign-up, sign-out process, and also shows the messages that are displayed to the user.

Reply:
@LeonRabinovich Approved, because of time constraints.
Please make sure this PR is merged ASAP.

Reply:
@LeonRabinovich Hey, sorry for taking long with my review.
I can see the upside in having this notification bar on the layout for general cases, but in my opinion code duplication is not a real problem here since each page will have its own messages, and I do prefer displaying msgs next to relevant elements for the user's actions.
Also, with that approach you maybe avoid writing similar code in two files but eventually the browser loads (due to extends) this code each time a page is loaded - any page - even if no messages will be used in it.
I would have done it differently but since we're short on time I don't think it's that important, and this notification bar can be used for other cases if needed.